### DISCUSSION TITLE
is there any support for offline and online sync?

### DISCUSSION THREAD

**Discussion by kakzaki** [pocketbase/pocketbase#67] [Q&A]
hi guys, thanks for this great project,

I was wondering if it supports realtime database and offline online sync just like firestore?

thanks

**Comment by ganigeorgiev** [comment]
PocketBase has realtime subscriptions, but doesn't have offline/online sync - this is left for users to handle.

This is how the minimal realtime subscriptions look like using the official [JavaScript SDK](https://github.com/pocketbase/js-sdk):
```js
import PocketBase from 'pocketbase';

const client = new PocketBase('https://example.com');

...

// subscribe to changes in any record from the "demo" collection
client.Realtime.subscribe("demo", function (e) {
    console.log(e.action, e.record);
});
```

You could also check the [Realtime API docs](https://pocketbase.io/docs/api-realtime/) for more information.

_p.s. I'm moving the issue to discussions since the format is more suitable for that._

**Comment by pchasco** [comment]
Offline sync is the killer feature that Firebase and Realm have over the other second-tier BaaS like SupaBase. It is difficult to implement with a relational database because referential integrity between records in different tables must be maintained after a sync, which means a sync must be performed in a controlled manner with records synced in the correct order. Document databases such as MongoDB are easier to sync because each document stands on its own, more or less, and can be synchronized independently of any other documents.

If you can denormalize your data to the point where there are no dependencies between tables, or simply store the bulk of your data as JSON text next to a primary key with updated date columns then sync becomes much simpler to implement.

**Reply by rodydavis** [reply]
Would also be nice to sync to a client side SQlite DB too (sqlite wasm / included db).

Or maybe something like: https://pouchdb.com/

**Reply by pchasco** [reply]
You would need to sync a client-side database in order for sync to be useful. For Flutter apps you would use something on SQLite, such as Drift, or just vanilla SQFlite. You could also use ObjectBox if you prefer a non-SQLite solution.

PouchDB is designed to sync against a CouchDB cluster. If not using a CouchDB backend I don't think there is any advantage to using PouchDB over another technology like those I mentioned.

**Reply by rodydavis** [reply]
I helped contribute to Moor (now Drift) and am thinking about making a similar abstraction. client side SQlite cache to use with the sdk

**Reply by rodydavis** [reply]
I created an example with flutter using SQLite (Drift) client side that stores the collections and even supports full text search:

https://github.com/rodydavis/pocketbase_drift

**Reply by gevera** [reply]
Maybe RxDB ?

**Reply by pchasco** [reply]
> I created an example with flutter using SQLite (Drift) client side that stores the collections and even supports full text search:
> 
> https://github.com/rodydavis/pocketbase_drift

Awesome! How does it work?

**Reply by rodydavis** [reply]
It  has a local SQLite database on top of the pocketbase client that then syncs with the server. I'm working on improving it and PRs welcomed

**Reply by HEAVYPOLY** [reply]
@rodydavis this looks awesome!  It's just plug and play and works!?  Is it possible to use this with vanilla js (non flutter)?

**Reply by rodydavis** [reply]
It is plugin in play! And writing a bunch of tests for it. It will return the same responses as the api (including sorting,fields and filtering).

Right now it is just dart because i have a flutter project i am building it for. But would love to add TS support too.

My goal is to simply extend the sdk to add the support so a user would not have to care. In most cases it simply adds a fetch policy to the request (defaults to cache and network)

**Comment by Sun3** [comment]
Realtime syncing is great, but in mobile applications the device can be offline without internet connection depending on your location. When the device if not able to connect to internet the users today need to be able to still use the app. For example adding an entry.

I sure hope that offline sync will be considered, because it's a killer feature.

Thank you.

**Comment by rodydavis** [comment]
Here is a complete example of showing 2 way sync between a PocketBase server and SQLite client (via drift in dart).

https://github.com/rodydavis/drift-pocketbase-sync

This does not use a CRDT but instead relies on the global ordering of the utc timestamps as a single source of truth on the server and the same method as the PocketBase source code for generating IDs on the client.

Deletes are handled by a hook and adding all deletes to a collection that clients can query to delete locally. There is a scheduled job to clear old ones after X amount of days.

This is done via db hooks and records can be deleted on the client or admin console or using the sdk as normal.

Deletes locally are a soft delete and will hard delete when synced.

The client (for non view collections) can also have offline mutations that store when it is synced, fresh, and deleted.

Each collection keeps track of the last time it was synced and only fetches updates.

Fresh is an optimization to know when to create/update via the sdk and ignore if deleted before a sync occurs.

There is a client sdk generator that generates the dart SQLite code, classes and helpers. Even the indexes defined in the admin console will be generated in the local database and any triggers added too.

Sync service:
https://github.com/rodydavis/drift-pocketbase-sync/blob/main/client/lib/db/sync.dart

**Reply by alterhuman** [reply]
@rodydavis this looks great! What happens in case of conflict (2 users updating the same row while offline)?

Also, is this (mostly) as good as CRDT?

**Reply by rodydavis** [reply]
Right now it is last right win on individual records but they are ordered universally through UTC. At least for my Dart client I had it made time stamps the same way the server did.

I am working on a partial updates for records that would be more inline with a CRDT. The biggest thing I have been having to account for is file uploads and how to track them. Right now I only support json type updates.

**Reply by leomorpho** [reply]
I'd definitely be interested in a ts/js version if you ever need that :)
